src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,1055 @@
+                raise ValueError(
+                    f"Mismatched inputs: nqubits given as {nqubits}. {class_qubit_gate} is a {qb_gate}-qubit gate."
+                )
+
+    # GST for empty circuit or operators
+    if nqubits is not None:
+        if backend is None:  # pragma: no cover
+            backend = GlobalBackend()
+
+        if nqubits == 1:
+            matrix_jk = np.zeros((4, 4))
+            for k in range(0, 4):
+                # Prepare state using "prepare_1qb_state(k)"
+                circ = prepare_states(k, nqubits)
+
+                # If class_qubit_gate is present (with/without theta)
+                if class_qubit_gate is not None:
+                    if theta is None:
+                        circ.add(getattr(gates, class_qubit_gate)(0))
+                    elif theta is not None:
+                        circ.add(getattr(gates, class_qubit_gate)(0, theta))
+
+                # Measurements
+                for j in range(0, 4):
+                    new_circ = measurement_basis(j, circ, nqubits)
+                    if noise_model is not None and backend.name != "qibolab":
+                        new_circ = noise_model.apply(new_circ)
+
+                    expectation_val = GST_execute_circuit(
+                        new_circ, k, j, nshots, nqubits
+                    )
+                    matrix_jk[j, k] = expectation_val
+
+            return matrix_jk
+
+        elif nqubits == 2:
+            matrix_jk = np.zeros((16, 16))
+            for k in range(0, 16):
+                # Prepare states using function "prepare_2qb_states(k)".
+                circ = prepare_states(k, nqubits)
+
+                # If class_qubit_gate is present (with/without theta).
+                if class_qubit_gate is not None:
+                    if theta is None:
+                        circ.add(
+                            getattr(gates, class_qubit_gate)(targ_qb[0], ctrl_qb[0])
+                        )
+                    else:
+                        circ.add(
+                            getattr(gates, class_qubit_gate)(
+                                targ_qb[0], ctrl_qb[0], theta
+                            )
+                        )
+
+                # Measurements
+                for j in range(0, 16):
+                    new_circ = measurement_basis(j, circ, nqubits)
+                    if noise_model is not None and backend.name != "qibolab":
+                        new_circ = noise_model.apply(new_circ)
+
+                    expectation_val = GST_execute_circuit(
+                        new_circ, k, j, nshots, nqubits
+                    )
+                    matrix_jk[j, k] = expectation_val
+
+            return matrix_jk
+
+        else:
+            raise ValueError(
+                f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+            )
+
+    else:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
```

Comment:
```suggestion
def GST(
    nqubits=None,
    gate=None,
    nshots=int(1e4),
    noise_model=None,
    backend=None,
):
    """Runs gate set tomography for a single qubit gate/circuit.

        Args:
        nshots (int, optional): Number of shots used in Gate Set Tomography.
        noise_model (:class:`qibo.noise.NoiseModel`, optional): Noise model applied
            to simulate noisy computation.
        save_data: To save gate set tomography data. If None, skip.
        class_qubit_gate: Identifier for the type of gate given in circuit.raw. If None,
            do gate set tomography with empty circuit.
        theta: Identifier for the gate parameter given in circuit.raw. If None, do gate
            set tomography with only gate provided by class_qubit_gate.
        backend (:class:`qibo.backends.abstract.Backend`, optional): Calculation engine.

    Returns:
        numpy.matrix: Matrix with elements jk equivalent to either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
    """

    # Check if class_qubit_gate corresponds to nqubits-gate.

    if nqubits != 1 and nqubits != 2:
        raise ValueError(
            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
        )
    
    if backend is None:  # pragma: no cover
        backend = GlobalBackend()

    if gate is not None:
        qb_gate = len(gate.qubits)
        if nqubits != qb_gate:
            raise ValueError(
                f"Mismatched inputs: nqubits given as {nqubits}. {gate} is a {qb_gate}-qubit gate."
            )
        gate = gate.__class__(*list(range(qb_gate)), **gate.init_kwargs)

    # GST for empty circuit or operators

    matrix_jk = np.zeros((4**nqubits, 4**nqubits))
    for k in range(4**nqubits):
        circ = prepare_states(k, nqubits)

        # If class_qubit_gate is present (with/without theta)
        if gate is not None:
            circ.add(gate)

        for j in range(4**nqubits):
            new_circ = measurement_basis(j, circ)

            if noise_model is not None and backend.name != "qibolab":
                new_circ = noise_model.apply(new_circ)

            expectation_val = GST_execute_circuit(
                new_circ, j, nshots, backend=backend
            )
            matrix_jk[j, k] = expectation_val

    return matrix_jk
```

Here it is possible to unify the single and two qubits cases. I think it's better to directly pass the gate instead of the `class_qubit_gate identifier`. This way, we can directly access the `theta` parameter. We just have to be careful to ensure that it acts on qubit 0 or on qubits 0 and 1. This makes the code shorter.

Reply:
Hi @AlejandroSopena , I was thinking of using ```circuit.raw['queue'][i]``` to extract the ```i```th gate. In your suggestion, what should ```gate``` be?